information/specification.md

```diff
@@ -0,0 +1,826 @@
+    algorithm has removed values to pay for entries in the [requested output
+    list](#requested-output-list).
+
+    In the context of a _wallet_, if a coin selection algorithm is applied to
+    the wallet's _complete_ UTxO set, then the _remaining_ UTxO set represents
+    the _updated_ UTxO set of that wallet.
+
+## Properties
+
+### Coverage of Payments
+
+This property states that the total value of the [coin
+selection](#coin-selection) result is sufficient to _cover_ the total value of
+the [requested output list](#requested-output-list).
+
+In particular:
+
+  * V<sub>_selected_</sub> ≥ V<sub>_requested_</sub>
+
+Where:
+
+  * V<sub>_requested_</sub>
+
+    is the total value of the [requested output list](#requested-output-list)
+
+
+  * V<sub>_selected_</sub>
+
+    is the total value of the _inputs_ field of the [coin
+    selection](#coin-selection) result.
+
+### Correctness of Change
+
+This property states that the correct amount of _change_ was generated.
+
+In particular:
+
+  * V<sub>_selected_</sub> = V<sub>_requested_</sub> + V<sub>_change_</sub>
+
+Where:
+
+  * V<sub>_change_</sub>
+
+    is the total value of the _change_ field of the [coin
+    selection](#coin-selection) result.
+
+  * V<sub>_requested_</sub>
+
+    is the total value of the [requested output list](#requested-output-list)
+
+  * V<sub>_selected_</sub>
+
+    is the total value of the _inputs_ field of the [coin
+    selection](#coin-selection) result.
+
+### Conservation of UTxO
+
+This property states that every entry in the [initial UTxO
+set](#initial-utxo-set) is included in _either_ the [coin
+selection](#coin-selection) inputs, _or_ in the [remaining UTxO
+set](#remaining-utxo-set), but _not both_.
+
+  * If a UTxO entry _is_ selected by the coin selection algorithm, it is
+    included in the [coin selection](#coin-selection) inputs.
+
+  * If a UTxO entry is _not_ selected by the coin selection algorithm, it is
+    included in the [remaining UTxO set](#remaining-utxo-set).
+
+The following laws hold:
+
+  * UTxO<sub>_selected_</sub> ⊆ UTxO<sub>_initial_</sub>
+  * UTxO<sub>_remaining_</sub> ⊂ UTxO<sub>_initial_</sub>
+  * UTxO<sub>_remaining_</sub> ∩ UTxO<sub>_selected_</sub> = ∅
+  * UTxO<sub>_remaining_</sub> ⋃ UTxO<sub>_selected_</sub> =
+    UTxO<sub>_initial_</sub>
+
+Where:
+
+  * **UTxO<sub>_initial_</sub>**
+
+    is the [initial UTxO set](#initial-utxo-set).
+
+  * **UTxO<sub>_remaining_</sub>**
+
+    is the [remaining UTxO set](#remaining-utxo-set).
+
+  * **UTxO<sub>_selected_</sub>**
+
+    is the value of the _inputs_ field of the [coin selection](#coin-selection)
+    result.
+
+### Conservation of Outputs
+
+This property states that the [requested output list](#requested-output-list)
+is _conserved_ in the [coin selection](#coin-selection) result.
+
+In particular, the _outputs_ field of the [coin selection](#coin-selection)
+result should be _equal to_ the [requested output list](#requested-output-list).
+
+## Failure Modes
```

Comment:
It feels odd to have both the possible failures and the properties described before even talking about the algorithms themselves. What about changing the order of the sections here so that we first describe the algorithms, and then, talk about the properties they have. 

It's like going through the troubleshooting section of an installation manual directly without even having looked at the installation instructions in the first place.

Reply:
> It feels odd to have both the possible failures and the properties described before even talking about the algorithms themselves. What about changing the order of the sections here so that we first describe the algorithms, and then, talk about the properties they have.

Hmm. I'm not sure I agree here.

I would argue that the set of potential failure modes is actually part of the API, as it represents a set of errors common to all algorithms. The API itself defines the set of potential errors. (This is also the case in the Haskell library.)

Similarly, the properties described in the properties section apply to all possible algorithms (and not just those we have provided). They specify additional restrictions on the relationships between parameters and results that are not practical to express in the types themselves.

The separation between API and implementation actually mirrors the separation we have in the library itself.